### DISCUSSION TITLE
Extensions: allow plugins to register links in panel right-click menu

### DISCUSSION THREAD

**Discussion by sd2k** [grafana/grafana#66037] [Plugins]
As part of https://github.com/grafana/grafana/issues/61658 it would be very cool if plugins could register actions on the right-click (or cmd/alt+click, or even cmd/alt+click+drag) menu of either a dashboard time-series panel or the Explore time-series panel.

Specifically, when the user right-clicks on the panel, a context menu would appear containing all registered links (as well as any core actions such as 'add annotation'). Similarly if the user cmd+clicks+drags a region on the panel.

The extension link's context could then contain several interesting things:
- the data queries in the panel
- the current time range
- the highlighted region time range, if the user highlighted a region
- the field info of the series that the user clicked on, if they clicked on a specific series
- possibly even the data in the panel, if it's available?

I manually hacked in something a bit like this for a hackathon project so it may be vaguely helpful (see [the uPlot plugin](https://github.com/grafana/grafana/blob/hackathon-2022-08-correographers/public/app/plugins/panel/timeseries/plugins/CorrelationsPlugin.tsx) and [its use in the TimeSeriesPanel](https://github.com/grafana/grafana/blob/a6d4a8e14d066daa910b51fb4469d4194fcd2ec3/public/app/plugins/panel/timeseries/TimeSeriesPanel.tsx#L93-L131), but that was before plugin extensions were a thing, so it used a similar technique to annotations (specifically: a datasource could [declare itself capable of running correlations](https://github.com/grafana/grafana/blob/a6d4a8e14d066daa910b51fb4469d4194fcd2ec3/packages/grafana-data/src/types/datasource.ts#L127), the panel would [display a row in the context menu for each such registered datasource instance](https://github.com/grafana/grafana/blob/a6d4a8e14d066daa910b51fb4469d4194fcd2ec3/public/app/plugins/panel/timeseries/plugins/CorrelationsPlugin.tsx#L282-L294), and [on click the Explore view would be split](https://github.com/grafana/grafana/blob/a6d4a8e14d066daa910b51fb4469d4194fcd2ec3/public/app/plugins/panel/timeseries/plugins/CorrelationsPlugin.tsx#L255-L259C10) using the query returned by a call to a new function in the datasource's API, which was [passed some info about the clicked/highlighted region](https://github.com/grafana/grafana/blob/a6d4a8e14d066daa910b51fb4469d4194fcd2ec3/public/app/plugins/panel/timeseries/plugins/CorrelationsPlugin.tsx#L244-L250)). I think it could be a lot cleaner with plugin extensions though!

**Comment by ifrost** [comment]
Explore exposes now `ExploreToolbarAction` extension point added in #65524. It's placed in "Add to..." dropdown in the top toolbar. The context contains all queries and results. Maybe that could be useful in this case?

**Reply by sd2k** [reply]
It's definitely an improvement! I'd still love to see more actions in the panel right click/drag menu though, I think that's an underused area of Grafana's viz.

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). We are migrating this discussion to an issue and closing the discussion. The issue is #82580. Feel free to continue the discussion around this there. Thank you!
